File: 2026/04-April/07-Tuesday/Solution.py
class Robot:

    def __init__(self, width: int, height: int):
        self.w = width
        self.h = height
        self.x = 0
        self.y = 0
        self.dir = 0  # 0=East, 1=North, 2=West, 3=South 
        self.cycle = 2 * (width + height) - 4
        self.moved = False

    def step(self, num: int) -> None:
        num %= self.cycle
        if num == 0 and self.moved:
            num = self.cycle

        self.moved = True

        for _ in range(num):
            if self.dir == 0:  # East
                if self.x + 1 < self.w:
                    self.x += 1
                else:
                    self.dir = 1
                    self.y += 1

            elif self.dir == 1:  # North
                if self.y + 1 < self.h:
                    self.y += 1
                else:
                    self.dir = 2
                    self.x -= 1

            elif self.dir == 2:  # West
                if self.x - 1 >= 0:
                    self.x -= 1
                else:
                    self.dir = 3
                    self.y -= 1

            else:  # South
                if self.y - 1 >= 0:
                    self.y -= 1
                else:
                    self.dir = 0
                    self.x += 1

    def getPos(self):
        return [self.x, self.y]

    def getDir(self):
        return ["East", "North", "West", "South"][self.dir]

File: 2026/04-April/07-Tuesday/Solution.java
class Robot {
    int w, h, x, y, dir;
    int cycle;
    boolean moved;

    public Robot(int width, int height) { 
        this.w = width;
        this.h = height;
        this.x = 0;
        this.y = 0;
        this.dir = 0; // 0=East, 1=North, 2=West, 3=South
        this.cycle = 2 * (width + height) - 4;
        this.moved = false;
    }

    public void step(int num) {
        num %= cycle;
        if (num == 0 && moved) {
            num = cycle;
        }

        moved = true;

        for (int i = 0; i < num; i++) {
            if (dir == 0) { // East
                if (x + 1 < w) {
                    x++;
                } else {
                    dir = 1;
                    y++;
                }
            } else if (dir == 1) { // North
                if (y + 1 < h) {
                    y++;
                } else {
                    dir = 2;
                    x--;
                }
            } else if (dir == 2) { // West
                if (x - 1 >= 0) {
                    x--;
                } else {
                    dir = 3;
                    y--;
                }
            } else { // South
                if (y - 1 >= 0) {
                    y--;
                } else {
                    dir = 0;
                    x++;
                }
            }
        }
    }

    public int[] getPos() {
        return new int[]{x, y};
    }

    public String getDir() {
        String[] dirs = {"East", "North", "West", "South"};
        return dirs[dir];
    }
}
